src/source.cpp

```diff
@@ -430,15 +517,30 @@ MeshSource::MeshSource(pugi::xml_node node)
 
 SourceSite MeshSource::sample(uint64_t* seed) const
 {
-  // sample location and element from mesh
-  auto mesh_location = space_->sample_mesh(seed);
-
-  // Sample source for the chosen element
-  int32_t element = mesh_location.first;
-  SourceSite site = source(element)->sample(seed);
+  // Sample the CDF defined in initialization above
+  int32_t element = space_->sample_element_index(seed);
+
+  // Sample position and apply rejection on spatial domains
+  Position r;
+  while (true) {
```

Comment:
Rather than a `while/break`, I'd suggest hearkening back to our Fortran roots here with the `do/while` construct more commonly used in that language.